KAFKA-5121: Implement transaction index for KIP-98

hachikuji opened the pull request.

hachikuji committed:
KAFKA-5121 [WIP]: Implement transaction index for KIP-98

ijuma commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -437,11 +438,21 @@ class GroupMetadataManager(val brokerId: Int,
-          val bufferRead = fileRecords.readInto(buffer, 0)
+          // We only add the READ_UNCOMMITTED parameter here because EasyMock in the GroupMetadataManagerTest
+          // cannot seem to deal with the default value for isolationLevel in log.read() -- it throws an
+          // IllegalStateException.
```
Seems like passing `isolationLevel` explicitly is a good idea anyway. Maybe we don't need this comment.

hachikuji replied:
Ack. Was thinking the same thing when I was looking over this.

ijuma commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -437,11 +438,21 @@ class GroupMetadataManager(val brokerId: Int,
+            isolationLevel = IsolationLevel.READ_UNCOMMITTED)
+
+          val memRecords = fetchDataInfo.records match {
+            case records: MemoryRecords => records
```
If we don't use `buffer` in this path, do we still want to allocate it eagerly?

hachikuji replied:
Sure, we can push allocation into the other branch. This is technically a bug fix, by the way, since `Log.read` can return `MemoryRecords.EMPTY`.

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -575,6 +618,20 @@ class Log(@volatile var dir: File,
   }
 
+  private def updateFirstUnstableOffset(): Unit = {
+    this.firstUnstableOffset = pidMap.firstUnstableOffset match {
```
You could use `map` instead of pattern matching.

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -735,38 +812,57 @@ class Log(@volatile var dir: File,
+            FetchDataInfo(fetchOffsetMetadata = fetchInfo.fetchOffsetMetadata,
+              records = fetchInfo.records,
+              firstEntryIncomplete = fetchInfo.firstEntryIncomplete,
+              abortedTransactions = Some(abortedTransactions))
```
Maybe this block should be in a separate method.

ijuma commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  private def openChannel(): FileChannel = {
+    val channel = new RandomAccessFile(file, "rw").getChannel
```
Nit: just use `FileChannel.open` directly?

ijuma commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,197 @@
+    buffer.putLong(firstOffset)
+    buffer.putLong(lastOffset)
+    buffer.putLong(lastStableOffset)
+    buffer.flip()
```
How do we go about deciding whether to use Schema classes or writing to the buffer directly?

hachikuji replied:
Kind of a subjective call, but since we scan the abort index when handling fetches and the schema is so simple, I thought we could skip the need to go through the `Struct` object when reading and writing.

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -436,8 +466,8 @@ class Log(@volatile var dir: File,
     */
-  def appendAsLeader(records: MemoryRecords, leaderEpoch: Int): LogAppendInfo = {
-    append(records, assignOffsets = true, leaderEpoch)
+  def appendAsLeader(records: MemoryRecords, leaderEpoch: Int, isFromClient: Boolean = true): LogAppendInfo = {
```
A general comment about default values: we should consider carefully when to use them as they are a common source of bugs. When we use them, the compiler no longer ensures that we think about what the value should be.

hachikuji replied:
Yeah, it was a tough call. There's only a handful of use cases for `isFromClient = false`, and the default is the more restrictive option, so I thought it was reasonable.

ijuma reviewed (commented):
A few random comments from a quick look. :)

hachikuji committed:
A few minor improvements from review

ijuma commented:
retest this please

hachikuji commented:
I realized that the relationship between the LSO and the high watermark is trickier than I originally thought. I had assumed that it was sufficient to take the minimum of the two as the maximum offset that a client could fetch in READ_COMMITTED mode. The issue is when the COMMIT or ABORT marker itself is larger than the high watermark. We need to ensure that none of the records from the associated transaction becomes readable until the marker itself is replicated (even if the records have offsets lower than the high watermark). This requires a little more bookkeeping as we need to keep track of which transactions have been completed, but are not replicated. I'll update the patch as soon as I have a solution ready.

hachikuji committed:
Ensure transactional data is visible only after txn marker has been replicated

hachikuji committed:
A few minor cleanups

ijuma commented:
retest this please

hachikuji commented:
note to self: once the transactional client patch lands, we should update `Fetcher` to use the control sequence number when finding the commit/abort markers.

hachikuji committed:
Fix a few problems after rebasing

hachikuji committed:
Support reading transaction index with DumpLogSegments

hachikuji committed:
Fix a few obvious problems

hachikuji committed:
Ensure LogValidator preserves transactional flag

hachikuji committed:
Ensure transaction state is rebuilt on log initialization and truncation

hachikuji committed:
Use control sequence number on client to detect commit markers

junrao commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,207 @@
+  if (file.exists)
+    openChannel()
+
+  def append(abortedTxn: AbortedTxn): Unit = Utils.writeFully(channel, abortedTxn.buffer.duplicate())
```
During append, perhaps it's useful to assert that abortedTxn are inserted in ascending lastOffset order?

junrao commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,207 @@
+              s"version ${AbortedTxn.CurrentVersion}")
+
+          if (abortedTxn.lastOffset >= offset) {
+            channel.truncate(position)
```
We don't set the position of channel after channel.truncate(), which seems ok. For consistency, we probably want to do the same in FileRecords.truncateTo()?

junrao commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,207 @@
+        var position = 0
+        val lastPosition = channel.position()
+        while (lastPosition - position >= AbortedTxn.TotalSize) {
+          val buffer = ByteBuffer.allocate(AbortedTxn.TotalSize)
```
It seems that we can just allocate a single buffer and reuse, instead of reallocating?

junrao commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,207 @@
+
+          val abortedTxn = new AbortedTxn(buffer)
+          if (abortedTxn.version > AbortedTxn.CurrentVersion)
+            trace(s"Reading aborted txn entry with version ${abortedTxn.version} as current " +
```
We throw an exception in the iterator below. Should we do the same thing here?

junrao commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,207 @@
+
+private[log] case class TxnIndexSearchResult(abortedTransactions: List[AbortedTransaction], isComplete: Boolean)
+
+class TransactionIndex(var file: File) extends Logging {
```
Could we add some comments on what's stored in this index?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -74,8 +80,20 @@ case class LogAppendInfo(var firstOffset: Long,
                          isDuplicate: Boolean = false)
 
+case class ControlRecord(controlType: ControlRecordType, pid: Long, epoch: Short, offset: Long, timestamp: Long)
+case class CompletedTxn(pid: Long, firstOffset: Long, lastOffset: Long, isAborted: Boolean) extends Ordered[CompletedTxn] {
```
It would be useful to document whether firstOffset and lastOffset are inclusive or exclusive.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1383,10 +1392,42 @@ class KafkaApis(val requestChannel: RequestChannel,
+        partition -> MemoryRecords.withControlRecord(controlRecordType, producerId, marker.producerEpoch)
+      }.toMap
 
+      replicaManager.appendRecords(
```
It seems that we are missing the logic to fence off the request based on coordinatorEpoch?

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1383,10 +1392,42 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }.toMap
 
+      replicaManager.appendRecords(
+        timeout = 30000, // FIXME: Add this to the request type? Or broker config?
```
We can probably use requestTimeoutMs.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -74,8 +80,20 @@ case class LogAppendInfo(var firstOffset: Long,
                          validBytes: Int,
                          offsetsMonotonic: Boolean,
                          producerAppendInfos: Map[Long, ProducerAppendInfo],
+                         completedTransactions: List[CompletedTxn] = null,
```
Could we document the new field in the java doc above?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -436,8 +456,8 @@ class Log(@volatile var dir: File,
     */
-  def appendAsLeader(records: MemoryRecords, leaderEpoch: Int): LogAppendInfo = {
-    append(records, assignOffsets = true, leaderEpoch)
+  def appendAsLeader(records: MemoryRecords, leaderEpoch: Int, isFromClient: Boolean = true): LogAppendInfo = {
```
Could we add the new field to the comment above? Ditto for line 485.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -74,8 +80,20 @@ case class LogAppendInfo(var firstOffset: Long,
+                         completedTransactions: List[CompletedTxn] = null,
                          isDuplicate: Boolean = false)
 
+case class ControlRecord(controlType: ControlRecordType, pid: Long, epoch: Short, offset: Long, timestamp: Long)
```
Now that we have a few different types of epoch, could we change epoch to producerEpoch?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -661,36 +712,70 @@ class Log(@volatile var dir: File,
+
+    if (batch.hasProducerId) {
+      val pid = batch.producerId
+      val maybeControlRecord = if (batch.baseSequence == RecordBatch.CONTROL_SEQUENCE) {
```
Could we add a method in RecordBatch to indicate whether the batch is a controlled batch or not? Also, it seems that we use a special sequence number to indicate whether a batch is controlled or not. Would it be better to use another bit in the attribute for that?

hachikuji committed:
Fix initialization of ProducerAppendInfo when loading from the log

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -25,17 +25,19 @@ import kafka.utils.{Logging, nonthreadsafe}
 }
 
-private[log] case class ProducerIdEntry(epoch: Short, lastSeq: Int, lastOffset: Long, offsetDelta: Int, timestamp: Long) {
+private[log] case class ProducerIdEntry(pid: Long, epoch: Short, lastSeq: Int, lastOffset: Long, offsetDelta: Int,
```
epoch => producerEpoch? Ditto in line 54.

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -54,48 +56,106 @@ private[log] class ProducerAppendInfo(val pid: Long, initialEntry: ProducerIdEnt
   private var lastOffset = initialEntry.lastOffset
   private var maxTimestamp = initialEntry.timestamp
+  private var currentTxnFirstOffset = initialEntry.currentTxnFirstOffset
+  private val startedTxns = ListBuffer.empty[OngoingTxn]
```
There can only be one opening transaction from a pid, right? If so, does startedTxns need to be a set?

hachikuji replied:
Hmm.. yes that sounds right.

hachikuji replied:
I realized that we need all started transactions in order to ensure that the LSO is updated correctly. We take all the started transactions from an append, add them to a sorted collection, and then remove the completed transactions in order of completion. This ensures that the LSO is always correct.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1383,10 +1392,42 @@ class KafkaApis(val requestChannel: RequestChannel,
+        internalTopicsAllowed = true,
+        isFromClient = false,
+        entriesPerPartition = controlRecords,
+        sendResponseCallback(producerId))
```
Hmm, it seems because of this, we could be sending multiple responses for the same WriteTxnMarkersRequest, which won't be right?

hachikuji replied:
The responses are accumulated and only sent after all callbacks have been received. I can change the name to clarify this.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -369,51 +371,71 @@ public static MemoryRecords withRecords(CompressionType compressionType, SimpleR
+        MemoryRecordsBuilder builder = builder(buffer, RecordBatch.CURRENT_MAGIC_VALUE, CompressionType.NONE,
+                TimestampType.CREATE_TIME, 0L, RecordBatch.NO_TIMESTAMP, pid, epoch, RecordBatch.CONTROL_SEQUENCE, true,
+                RecordBatch.NO_PARTITION_LEADER_EPOCH);
+        builder.appendControlRecord(System.currentTimeMillis(), type, null);
```
The spec says the value of the control record contains the coordinatorEpoch.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -369,51 +371,71 @@ public static MemoryRecords withRecords(CompressionType compressionType, SimpleR
     }
 
+    public static MemoryRecords withControlRecord(ControlRecordType type, long pid, short epoch) {
+        ByteBuffer buffer = ByteBuffer.allocate(128);
```
Could we be calculate the buffer size more precisely?

hachikuji committed:
squash

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -661,36 +712,70 @@ class Log(@volatile var dir: File,
+
+      maybeControlRecord match {
+        case Some(controlRecord) =>
+          val completedTxn = appendInfo.appendControl(controlRecord)
```
Hmm, when this is called on the leader side, we haven't assigned the offset for the record yet. It seems that this means the offset returned in appendInfo.completedTransactions will be incorrect when we try to use it?

junrao commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -54,48 +56,106 @@ private[log] class ProducerAppendInfo(val pid: Long, initialEntry: ProducerIdEnt
-    append(entry.epoch, entry.firstSeq, entry.lastSeq, entry.timestamp, entry.lastOffset)
+        currentTxnFirstOffset = -1
+        maxTimestamp = controlRecord.timestamp
+        CompletedTxn(pid, firstOffset, lastOffset, controlRecord.controlType == ControlRecordType.ABORT)
```
When firstOffset is set to controlRecord.offset, it seems that lastOffset could be smaller than firstOffset. Should we guard that?

hachikuji replied:
I've fixed this. We always use the offset of the control record as the last offset.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -549,19 +569,31 @@ class Log(@volatile var dir: File,
         updateLogEndOffset(appendInfo.lastOffset + 1)
 
+        // update the first unstable offset (which is used to compute LSO)
+        updateFirstUnstableOffset()
```
Do we need to call this here given that we are calling the same method in onHighWatermarkIncremented()?

hachikuji replied:
Hmm... good point. I had thought it was necessary since the first unstable offset could be lower than the high watermark, but we wouldn't be able to advance it any further from this write because any appended transaction markers could not have been replicated yet.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -725,7 +812,9 @@ class Log(@volatile var dir: File,
-  def read(startOffset: Long, maxLength: Int, maxOffset: Option[Long] = None, minOneMessage: Boolean = false): FetchDataInfo = {
+  def read(startOffset: Long, maxLength: Int, maxOffset: Option[Long] = None,
+           minOneMessage: Boolean = false,
+           isolationLevel: IsolationLevel = IsolationLevel.READ_UNCOMMITTED): FetchDataInfo = {
```
Could we add isolationLevel to the comment above?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -776,6 +870,40 @@ class Log(@volatile var dir: File,
+  private def addAbortedTransactions(startOffset: Long, segmentEntry: JEntry[JLong, LogSegment],
+                                     fetchInfo: FetchDataInfo): FetchDataInfo = {
+    val fetchSize = fetchInfo.records.sizeInBytes
+    val startOffsetPosition = OffsetPosition(startOffset, fetchInfo.fetchOffsetMetadata.relativePositionInSegment)
```
We probably want to use the offset in fetchInfo instead of startOffset, which can be larger than startOffset?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1132,6 +1260,7 @@ class Log(@volatile var dir: File,
 
       pidMap.truncate()
       pidMap.updateMapEndOffset(newOffset)
+      updateFirstUnstableOffset()
```
Should we call updateFirstUnstableOffset() in truncateTo() too?

hachikuji replied:
I think we already do. It is in the call to `buildAndRecoverPidMap`.

junrao reviewed (commented):
@hachikuji : Thanks for the patch. Made a pass on the server side files and left some comments.

hachikuji committed:
Fix some breakage after rebasing

hachikuji committed:
Switch to read/write lock and address a few miscellaneous comments

hachikuji committed:
Use correct offset information when building producer append info

hachikuji committed:
Add more test cases

hachikuji committed:
Fix broken test case

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -575,6 +627,25 @@ class Log(@volatile var dir: File,
+        firstUnstableOffset
+      } else {
+        val segment = segments.floorEntry(offset).getValue
+        val position  = segment.translateOffset(offset)
```
Getting the position requires index lookup and may be a bit expensive. Do we really need to maintain position? It seems that when using firstUnstableOffset, we only need the offset, not the position?

junrao commented on core/src/main/scala/kafka/cluster/Replica.scala:
```diff
@@ -143,6 +144,23 @@ class Replica(val brokerId: Int,
+  def lastStableOffset: LogOffsetMetadata = {
+    log.map { log =>
+      log.firstUnstableOffset match {
+        case Some(offsetMetadata) if offsetMetadata.messageOffset < highWatermark.messageOffset => offsetMetadata
```
Hmm, log.firstUnstableOffset is only advanced after the completeMarker has been fully replicated, which suggests that log.firstUnstableOffset is always going to be < highWatermark?

hachikuji committed:
Implement transaction index recovery

hachikuji commented:
@junrao I've added a commit to do transaction index recovery. The approach I've taken is the following:

1. I've removed the periodic PID snapshot.
2. Instead, when the log rolls, I take a new snapshot using the offset of the new segment.
3. When a segment has been flushed to disk, I remove the corresponding snapshot. The snapshot corresponding the start of the active segment is never removed.
4. When recovering a segment, I have to read the log from the starting segment in the worst case in order to rebuild the transaction index. Once finished rebuilding an individual segment, I take another snapshot so that we don't have to repeat the same work if we have to recover another segment. To make this efficient, I ensure that logs are loaded/recovered in ascending order.
5. When the log is cleanly closed, I take another snapshot. I was a bit unsure of the benefit of this. If we can always or usually expect to truncate upon startup, then perhaps the benefit is low and we need to revisit periodic PID expiration to reduce the chance that we have to scan the last segment upon initialization.

hachikuji committed:
Fix broken LSO update logic when replicating

hachikuji committed:
Fix failing test case after update to collect all started transactions

hachikuji committed:
Add test case for control record builder

hachikuji committed:
Remove lock in TransactionIndex to follow pattern similar to FileRecords

junrao commented on core/src/main/scala/kafka/log/TransactionIndex.scala:
```diff
@@ -0,0 +1,235 @@
+    iterator(() => ByteBuffer.allocate(AbortedTxn.TotalSize)).map(_._1).toList
+  }
+
+  def collectAbortedTxns(fetchOffset: Long, upperBoundOffset: Long): TxnIndexSearchResult = {
```
Could we add a comment whether fetchOffset and upperBoundOffset are inclusive or exclusive?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1007,6 +1176,10 @@ class Log(@volatile var dir: File,
       }
+
+      // take a snapshot of the producer states to facilitate recovery
+      producerStateManager.takeSnapshot()
```
When we call roll(), the base offset of the new segment could actually be larger than logEndOffset. So, in that case, the snapshot offset may not match the base offset of the next segment. Since in flush(), we delete snapshots based on the baseOffset of the active segment, we may delete the last snapshot by accident. One way to fix this is to call producerStateManager.updateMapEndOffset(newOffset) before taking the snapshot.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -74,9 +84,33 @@ case class LogAppendInfo(var firstOffset: Long,
+                        lastOffset: Long,
+                        isAborted: Boolean) extends Ordered[CompletedTxn] {
+  override def compare(that: CompletedTxn): Int = {
+    val res = this.firstOffset compare that.firstOffset
```
Hmm, the comment in line 72 says completed txns are sorted by last offset, which seems to be what we want?

hachikuji replied:
Sorting by the first offset is what we want. This is used to find the first unstable offset which will always be the minimum first offset of all transactions. I couldn't find the comment you were referring to since the patch was updated. If it still exists, can you point me to it?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -279,25 +306,45 @@ class Log(@volatile var dir: File,
               Files.deleteIfExists(timeIndexFile.toPath)
               Files.delete(indexFile.toPath)
-              segment.recover(config.maxMessageSize)
+              segment.txnIndex.delete()
```
The warn in line 312 now needs to include the tnx index.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -279,25 +306,45 @@ class Log(@volatile var dir: File,
-    // log files that are replaced by the swap segment should be renamed to .deleted
-    // before the swap file is restored as the new segment file.
+  private def recoverSegment(segment: LogSegment, leaderEpochCache: Option[LeaderEpochCache] = None): Int = {
+    val stateManager = new ProducerStateManager(config, topicPartition, dir, maxPidExpirationMs)
```
This can be done later. But reloading the snapshot when recovering every segment can be expensive. Since segments are recovered in order, it seems that we just need to load the snapshot on recovering the first segment.

hachikuji committed:
Idempotent/transactional validation should come after offset assignment while holding the Log lock

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -250,7 +275,21 @@ class LogSegment(val log: FileRecords,
+          if (batch.hasProducerId) {
+            val producerId = batch.producerId
+            if (batch.baseSequence == RecordBatch.CONTROL_SEQUENCE) {
+              val firstOffset = ongoingTransactions.remove(producerId).getOrElse(throw new IllegalStateException())
```
Hmm, in theory, it seems that we could have an abort marker with no open records before it?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -669,7 +668,8 @@ public void onFailure(RuntimeException e) {
                                                                                  boolean requireTimestamp) {
-        ListOffsetRequest.Builder builder = ListOffsetRequest.Builder.forConsumer(requireTimestamp)
+        ListOffsetRequest.Builder builder = ListOffsetRequest.Builder
+                .forConsumer(requireTimestamp, IsolationLevel.READ_UNCOMMITTED)
```
Hmm, instead of using hard-coded READ_UNCOMMITTED, shouldn't we use the isolation_level in the Fetcher?

hachikuji replied:
Ah, good catch. Forgot to update this after the transactional client patch was merged.

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -256,74 +384,73 @@ class ProducerIdMapping(val config: LogConfig,
    * Truncate the PID mapping to the given offset range and reload the entries from the most recent
-   * snapshot in range (if there is one).
+   * snapshot in range (if there is one). Note that the log end offset is assumed to be less than
+   * or equal to the high watermark.
```
Hmm, not sure about the comment on HW. With KIP-101, typically the truncation point is >= the local HW.

junrao reviewed (commented):
@hachikuji : Thanks for the updated patch. Looks good overall. A few more comments. The big remaining items are (1) fencing writeTxnMarker based on coordinator epoch; (2) we need to make a decision on whether to use a batch level attribute bit to represent txnEndMarker instead of special sequence number.

hachikuji committed:
Add test case for transaction index recovery

hachikuji committed:
Some cleanup for better code reuse during txn index recovery

hachikuji committed:
Duplicate check must be done under the Log lock

hachikuji committed:
Basic support for transaction coordinator fencing

hachikuji committed:
Add a few test cases to validate coordinator fencing

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -534,34 +593,50 @@ class Log(@volatile var dir: File,
+          appendInfo.lastOffset = duplicate.lastOffset
+          appendInfo.logAppendTime = duplicate.timestamp
+          return appendInfo
+        }
```
Hmm, this logic may not be quite right for the follower. The follower could append batches for different pids in a single append call, if we reject all batches because a single batch is duplicated, the follower will miss records from other pids.

In theory, duplicates should never happen in the follower. If duplicates somehow already leaked into the leader, it's kind of hard to not take the duplicates in the follower since the follower's log is supposed to be identical with the leader. So, perhaps we can only return on duplicates if the append is on the leader. If duplicates are detect in the follower, just log a warning but proceed with the append?

hachikuji replied:
The intent was to return a duplicate only if `isFromClient` is set (which would never be the case for the follower), but it looks like I left that out. I'll fix in the next commit.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -363,57 +377,89 @@ public static MemoryRecordsBuilder builder(ByteBuffer buffer,
+                TimestampType.CREATE_TIME, initialOffset, RecordBatch.NO_TIMESTAMP, producerId, producerEpoch,
+                RecordBatch.CONTROL_SEQUENCE, true, RecordBatch.NO_PARTITION_LEADER_EPOCH);
+        ByteBuffer value = marker.serializeValue();
+        builder.appendControlRecord(System.currentTimeMillis(), marker.controlType(), value);
```
It's a bit inconsistent that we serialize the value here, but the key inside appendControlRecord(). Perhaps it's better to serialize the value inside appendControlRecord() too?

hachikuji replied:
I was trying to keep `appendControlRecord` generic for future control record types. That said, it seems reasonable to add an `appendEndTxnMarker` which handles serialization of both the key and value.

junrao reviewed (commented):
A couple of more comments.

hachikuji committed:
Address a few review comments

hachikuji committed:
Ensure control records are not compressed in LogValidator

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -721,11 +813,19 @@ class Log(@volatile var dir: File,
    * @param minOneMessage If this is true, the first message will be returned even if it exceeds `maxLength` (if one exists)
+   * @param isolationLevel The isolation level of the fetcher. The READ_UNCOMMITTED isolation level has the traditional
+   *                       read semantics (e.g. consumers are limited to fetching up to the high watermark). In
+   *                       READ_COMMITTED, consumers are limited to fetching up the the last stable offset. Additionally,
```
fetching up the the  => fetching up to the

hachikuji committed:
Duplicates on follower append should cause error plus cleanup of EndTransactionMarker writing

hachikuji committed:
Allocate end transaction buffer size precisely

hachikuji committed:
Cache offset position after segment append for started transactions on leader

hachikuji committed:
Fix failing test cases

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -534,34 +583,56 @@ class Log(@volatile var dir: File,
 
-        // now append to the log
-        segment.append(firstOffset = appendInfo.firstOffset,
+        val segmentPosition = segment.append(firstOffset = appendInfo.firstOffset,
```
Do we need to get position from segment.append()? We could also just get the size of the segment before appending.

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -0,0 +1,589 @@
+    else if (undecidedFirstOffset.isEmpty)
+      unreplicatedFirstOffset
+    else if (undecidedFirstOffset.get.messageOffset < unreplicatedFirstOffset.get.messageOffset)
+      undecidedFirstOffset
```
It seems that undecidedFirstOffset should never be < unreplicatedFirstOffset?

hachikuji replied:
I think it is possible. You may have something like this:

W1, W2, C2, C1

(where W1 is a write from producer 1, and C1 is a commit from producer 1).

Say the high watermark has reached C1 (which means the transaction from producer 1 is not yet visible). In this case, the first undecided offset will be W1 while the first unreplicated offset will be W2.

hachikuji replied:
The example was a little wrong. The proper scenario is this:

W1, W2, C2

Say the high watermark is at C2. The first undecided offset will be W1, but the first unreplicated offset will be W2. Once the high watermark reaches C2+1, the first undecided offset will still be W1 and the first unreplicated offset will be empty.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -157,7 +169,7 @@ class Log(@volatile var dir: File,
     leaderEpochCache.clearEarliest(logStartOffset)
 
-    buildAndRecoverPidMap(logEndOffset)
+    loadProducerState(logEndOffset)
```
Do we need to do this under lock since we are updating the state in ProducerStateManager?

junrao commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -0,0 +1,589 @@
+  /**
+   * Returns the last offset of this map
+   */
+  def mapEndOffset = lastMapOffset
```
Are the callers all properly synchronized on Log.lock?

junrao reviewed (commented):
@hachikuji : Thanks for updating the patch. A few more comments on the latest fix.

hachikuji committed:
Remove control sequence sentinel and move control attribute to batch level

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -292,6 +292,16 @@ public static MemoryRecordsBuilder builder(ByteBuffer buffer,
+                                                         CompressionType compressionType,
+                                                         long baseOffset,
+                                                         long producerId,
+                                                         short epoch,
```
epoch => producerEpoch ?

junrao commented on core/src/main/scala/kafka/tools/DumpLogSegments.scala:
```diff
@@ -349,9 +360,14 @@ object DumpLogSegments {
+            if (controlType == ControlRecordType.ABORT || controlType == ControlRecordType.COMMIT) {
+              val endTxnMarker = EndTransactionMarker.deserialize(record)
+              print(s" coordinatorEpoch: ${endTxnMarker.coordinatorEpoch}")
+            }
```
Should we print out an error in the else clause?

hachikuji replied:
I added some code to print the unknown type id in the else case. Let me know if that seems reasonable.

hachikuji committed:
Retain snapshots from last two segments and cleanup locking producer state manager

hachikuji committed:
A few more cleanup/fixes

hachikuji committed:
Cleanup imports of LogOffsetTest

hachikuji committed:
Synchronize segment recovery in Log

junrao commented:
@hachikuji : Thanks for the patch. LGTM.

junrao commented:
I will let you merge this after running the system tests.

hachikuji commented:
@junrao Thanks for the review cycles. System tests are looking good: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-05-05--001.1494059152--hachikuji--eos-txn-index--64671d8/report.html. I am going to look at a couple of the build failures above before merging to see if they could be related.

hachikuji commented:
retest this please

hachikuji commented:
retest this please

hachikuji commented:
It looks like the recent transient failures are known issues: https://issues.apache.org/jira/browse/KAFKA-5183, https://issues.apache.org/jira/browse/KAFKA-5182, https://issues.apache.org/jira/browse/KAFKA-5175. In one of the previous builds, I saw a failure in kafka.server.LogRecoveryTest.testHWCheckpointWithFailuresSingleLogSegment. This could be related to this patch and there is no active JIRA, so I'm trying to reproduce it locally.

hachikuji committed:
Bump up timeout in waitUntilLeaderIsElectedOrChanged

hachikuji commented:
I wasn't able to reproduce the failure in `LogRecoveryTest` locally and it didn't occur in the last few builds. I went ahead and increased the timeout that failed just in case. I'm going to go ahead and merge this to trunk.
